version/version.go

```diff
@@ -5,98 +5,162 @@
-// of the form "1.23.0-dev20220316".
-var Short = ""
-
-func init() {
-	defer func() {
-		// Must be run after Short has been initialized, easiest way to do that
-		// is a defer.
-		majorMinorPatch, _, _ = strings.Cut(Short, "-")
-	}()
-
-	if Long != "" && Short != "" {
-		// Built in the recommended way, using build_dist.sh.
-		return
+// Stamp vars can have their value set at build time by linker flags (see
+// build_dist.sh for an example). When set, these stamps serve as additional
+// inputs to computing the binary's version as returned by the functions in this
+// package.
+//
+// All stamps are optional.
+var (
+	// longStamp is the full version identifier of the build. If set, it is
+	// returned verbatim by Long() and other functions that return Long()'s
+	// output.
+	longStamp string
+
+	// shortStamp is the short version identifier of the build. If set, it
+	// is returned verbatim by Short() and other functions that return Short()'s
+	// output.
+	shortStamp string
+
+	// gitCommitStamp is the git commit of the github.com/tailscale/tailscale
+	// repository at which Tailscale was built. Its format is the one returned
+	// by `git rev-parse <commit>`. If set, it is used instead of any git commit
+	// information embedded by the Go tool.
+	gitCommitStamp string
+
+	// gitDirtyStamp is whether the git checkout from which the code was built
+	// was dirty. Its value is ORed with the dirty bit embedded by the Go tool.
+	//
+	// We need this because when we build binaries from another repo that
+	// imports tailscale.com, the Go tool doesn't stamp any dirtiness info into
+	// the binary. Instead, we have to inject the dirty bit ourselves here.
+	gitDirtyStamp bool
+
+	// extraGitCommit, is the git commit of a "supplemental" repository at which
+	// Tailscale was built. Its format is the same as gitCommit.
+	//
+	// extraGitCommit is used to track the source revision when the main
+	// Tailscale repository is integrated into and built from another repository
+	// (for example, Tailscale's proprietary code, or the Android OSS
+	// repository). Together, gitCommit and extraGitCommit exactly describe what
+	// repositories and commits were used in a build.
+	extraGitCommitStamp string
+)
+
+var long lazy.SyncValue[string]
+
+// Long returns a full version number for this build, of one of the forms:
+//
+//   - "x.y.z-commithash-otherhash" for release builds distributed by Tailscale
+//   - "x.y.z-commithash" for release builds built with build_dist.sh
+//   - "x.y.z-changecount-commithash-otherhash" for untagged release branch
+//     builds by Tailscale (these are not distributed).
+//   - "x.y.z-changecount-commithash" for untagged release branch builds
+//     built with build_dist.sh
+//   - "x.y.z-devYYYYMMDD-commithash{,-dirty}" for builds made with plain "go
+//     build" or "go install"
+//   - "x.y.z-ERR-BuildInfo" for builds made by plain "go build" or "go install"
```

Comment:
Don't they? I thought it was "go run" that didn't.

Reply:
Ah, it's go run that result in that! Updated the docstring... And then also tested it and found that the version data for `go run` was all broken because build metadata is present, but doesn't contain a commit or build date. 

This was also broken before this PR, so I fixed that as a separate commit on top of the stack, as a plain standalone bugfix.